src/main/java/org/truffleruby/language/loader/RequireNode.java

```diff
@@ -80,34 +80,38 @@ private boolean requireWithMetrics(String feature, String expandedPathRaw, Dynam
-        if (autoloadConstants != null) {
-            if (getContext().getOptions().LOG_AUTOLOAD) {
-                String info = autoloadConstants
-                        .stream()
-                        .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
-                        .collect(Collectors.joining(" and "));
-                RubyLanguage.LOGGER
-                        .info(() -> String.format(
-                                "%s: requiring %s which is registered as an autoload for %s",
-                                RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
-                                feature,
-                                info));
-            }
+        try {
+            if (autoloadConstants != null) {
+                if (getContext().getOptions().LOG_AUTOLOAD) {
+                    String info = autoloadConstants
+                            .stream()
+                            .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
+                            .collect(Collectors.joining(" and "));
+                    RubyLanguage.LOGGER
+                            .info(() -> String.format(
+                                    "%s: requiring %s which is registered as an autoload for %s",
+                                    RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
+                                    feature,
+                                    info));
+                }
 
-            for (RubyConstant autoloadConstant : autoloadConstants) {
-                GetConstantNode.autoloadConstantStart(autoloadConstant);
-            }
-            try {
-                return doRequire(feature, expandedPath, pathString);
-            } finally {
                 for (RubyConstant autoloadConstant : autoloadConstants) {
-                    GetConstantNode.autoloadUndefineConstantIfStillAutoload(autoloadConstant);
-                    GetConstantNode.autoloadConstantStop(autoloadConstant);
-                    featureLoader.removeAutoload(autoloadConstant);
+                    GetConstantNode.autoloadConstantStart(autoloadConstant);
                 }
             }
-        } else {
+
             return doRequire(feature, expandedPath, pathString);
+        } finally {
+            final List<RubyConstant> releasedConstants = featureLoader.getAutoloadConstants(expandedPath);
+            if (releasedConstants != null) {
+                for (RubyConstant autoloadConstant : releasedConstants) {
+                    if (autoloadConstant.getAutoloadConstant().isAutoloading()) {
+                        GetConstantNode.autoloadUndefineConstantIfStillAutoload(autoloadConstant);
+                        GetConstantNode.autoloadConstantStop(autoloadConstant);
+                        featureLoader.removeAutoload(autoloadConstant);
+                    }
+                }
+            }
```

Comment:
So the change here is to call `featureLoader.getAutoloadConstants(expandedPath)` a second time, and merge the `else` branch since it's the same?

Could you explain why calling `featureLoader.getAutoloadConstants(expandedPath)` a second time helps, in what scenario? And add a comment in the code about that?